pretrain/tinyllama.py

```diff
@@ -0,0 +1,366 @@
+"""
+This script is adapted from TinyLlama:
+https://github.com/jzhang38/TinyLlama/blob/main/pretrain/tinyllama.py
+
+
+TODO LIST:
+- [x] check that seed is correctly set and each rank sees a partition of the data
+- [x] implement init-weights
+- [x] install torch nightly
+- [ ] use fake dataset to compare batches/sec numbers
+- [ ] determine global batch size
+- [ ] add torch.compile
```

Comment:
Here are some numbers

```python
import sys
import time
from contextlib import nullcontext
from datetime import datetime
from pathlib import Path

import torch
from torch.utils.data import DataLoader

import lightning as L
from lightning.fabric.strategies import FSDPStrategy

# support running without installing as a package
wd = Path(__file__).parent.parent.resolve()
sys.path.append(str(wd))

from lit_gpt import Config
from lit_gpt.model import GPT, Block
from lit_gpt.utils import num_parameters

model_name = "open_llama_3b"
name = "openwebtext-llama"
data_dir = Path("data") / name
log_interval = 1

# Hyperparameters
learning_rate = 6e-4
micro_batch_size = 2
max_iters = 25

hparams = {k: v for k, v in locals().items() if isinstance(v, (int, float, str)) and not k.startswith("_")}


def setup(compile: str = "eager", devices: int = 1, profile: bool = False) -> None:
    if devices > 1:
        strategy = FSDPStrategy(
            auto_wrap_policy={Block}, activation_checkpointing_policy={Block}, limit_all_gathers=True, cpu_offload=False
        )
    else:
        strategy = "auto"

    fabric = L.Fabric(devices=devices, strategy=strategy, precision="bf16-true")
    fabric.print(hparams)
    fabric.launch(main, compile, profile)


def main(fabric: L.Fabric, compile: str, profile: bool) -> None:
    fabric.seed_everything(1337, workers=True)  # same seed for every process to init model (FSDP)

    config = Config.from_name(model_name)
    assert config.block_size == 2048  # update FakeDataset otherwise
    fabric.print(f"Loading model with {config.__dict__}")
    t0 = time.perf_counter()
    with fabric.init_module(empty_init=True):
        model = GPT(config)
    fabric.print(f"Time to instantiate model: {time.perf_counter() - t0:.02f} seconds.")
    fabric.print(f"Total parameters {num_parameters(model):,}")

    if compile == "inductor":
        if isinstance(fabric.strategy, FSDPStrategy):
            # https://github.com/pytorch/pytorch/issues/101168
            model = torch.compile(model)
        else:
            model = torch.compile(model, fullgraph=True, mode="reduce-overhead")
    elif compile != "eager":
        raise ValueError(compile)

    model = fabric.setup(model)
    optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate, foreach=False)
    optimizer = fabric.setup_optimizers(optimizer)

    train_dataloader = DataLoader(FakeDataset(), batch_size=micro_batch_size, pin_memory=True)
    train_dataloader = fabric.setup_dataloaders(train_dataloader)

    train(fabric, model, optimizer, train_dataloader, profile)

    print(f"{torch.cuda.max_memory_allocated() / 1e9:.02f} GB")


def train(
    fabric: L.Fabric,
    model: torch.nn.Module,
    optimizer: torch.optim.Optimizer,
    train_dataloader: DataLoader,
    profile: bool,
) -> None:
    train_iter = iter(train_dataloader)
    for i in range(max_iters):
        iter_t0 = time.perf_counter()
        input_ids, targets = next(train_iter)

        if profile and i == max_iters - 1:
            prof = torch.profiler.profile(record_shapes=True, with_stack=True)
            torch.cuda.synchronize()
        else:
            prof = nullcontext()

        with prof:
            logits = model(input_ids)
            logits = logits.reshape(-1, logits.size(-1))
            targets = targets.reshape(-1)
            loss = torch.nn.functional.cross_entropy(logits, targets, ignore_index=-1)
            fabric.backward(loss)
            optimizer.step()
            optimizer.zero_grad()

        if hasattr(prof, "export_chrome_trace"):
            torch.cuda.synchronize()
            trace_dir = Path("traces/")
            trace_dir.mkdir(exist_ok=True)
            now = datetime.utcnow().isoformat().replace(" ", "_")
            trace_file = str(trace_dir / f"{compile}_{now}.json")
            print(f"Saving {trace_file}")
            prof.export_chrome_trace(trace_file)
        if i % log_interval == 0:
            loss_item = loss.item()
            t1 = time.perf_counter()
            fabric.print(f"iter {i}: loss {loss_item:.4f}, iter time: {(t1 - iter_t0) * 1000:.2f}ms")


class FakeDataset(torch.utils.data.Dataset):
    def __getitem__(self, index): return torch.randint(0, 10, size=(2048,)), torch.randint(0, 10, size=(2048,))
    def __len__(self): return 1000


if __name__ == "__main__":
    torch.set_float32_matmul_precision("high")

    from jsonargparse import CLI

    CLI(setup)


# eager 1
# iter 24: loss 10.3735, iter time: 567.70ms
# 26.92 GB

# eager 2
# iter 24: loss 8.7126, iter time: 794.05ms
# 9.01 GB

# inductor 1
# iter 24: loss 10.3735, iter time: 415.80ms
# 20.95 GB

# inductor 2
# iter 24: loss 8.7128, iter time: 588.75ms
# 9.06 GB

# PyTorch version: 2.2.0.dev20231030+cu121
# Is debug build: False
# CUDA used to build PyTorch: 12.1
# OS: Ubuntu 22.04.3 LTS (x86_64)
# Python version: 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] (64-bit runtime)
# Python platform: Linux-5.15.0-75-generic-x86_64-with-glibc2.35
# Is CUDA available: True
# CUDA runtime version: 12.1.105
# CUDA_MODULE_LOADING set to: LAZY
# GPU models and configuration: 
# GPU 0: NVIDIA A100-SXM4-40GB
# GPU 1: NVIDIA A100-SXM4-40GB
# Nvidia driver version: 525.125.06
```

Reply:
Hey @carmocca
Thanks for the numbers. 
You changed many conditions for the training script. For example, the optimizer, the precision, the loss function, the gradient accumulation etc. Then the numbers won't reflect what happens in the real training script. Could you help me understand the reasons for these changes?

Reply:
This is a completely different script that I use for benchmarking. I only meant to share the numbers and the corresponding code in case you wanted to try it